Lib/idlelib/debugger.py

```diff
@@ -216,142 +347,131 @@ def interaction(self, message, frame, info=None):
+        self.show_vars()
+        self.sync_source_line()
+        self.enable_buttons(self.buttons)
+
         self.top.wakeup()
         # Nested main loop: Tkinter's main loop is not reentrant, so use
         # Tcl's vwait facility, which reenters the event loop until an
         # event handler sets the variable we're waiting on
         self.nesting_level += 1
         self.root.tk.call('vwait', '::idledebugwait')
         self.nesting_level -= 1
-        #
-        for b in self.buttons:
-            b.configure(state="disabled")
-        self.status.configure(text="")
-        self.error.configure(text="", background=self.errorbg)
         self.frame = None
 
+    def show_vars(self):
+        self.vars.move(self.locals, '', 0)
+        self.vars.move(self.globals, '', 1)
+        self.vars.delete(*self.vars.get_children(self.locals))
+        self.vars.delete(*self.vars.get_children(self.globals))
+        self.var_values = {}
+        sel = self.stack.selection()
+        if len(sel) == 1 and sel[0] in self.framevars:
+            locals, globals, _, _ = self.framevars[sel[0]]
+            # note: locals/globals may be from a remotedebugger, in
+            # which case for reasons we don't need to get into here,
+            # they aren't iterable
+            self.add_varheader()
+            for name in sorted(locals.keys(), key=underscore_at_end):
+                self.add_var(name, locals[name])
+            self.add_varheader(isGlobal=True)
+            for name in sorted(globals.keys(), key=underscore_at_end):
+                self.add_var(name, globals[name], isGlobal=True)
+
+    def add_varheader(self, isGlobal=False):
+        pass
+
+    def add_var(self, varname, value, isGlobal=False):
+        item = self.vars.insert(self.globals if isGlobal else self.locals,
+                         'end', text=varname, values=(value, ))
+        self.var_values[item] = value
+
+    def mouse_moved_vars(self, ev):
+        # tooltip_schedule(ev, self.var_tooltip)
+        pass
+
+    def leave_vars(self, ev):
+        # tooltip_clear()
+        pass
+
+    def var_tooltip(self, ev):
+        # Callback from tooltip package to return text of tooltip
+        item = None
+        if self.vars.identify('column', ev.x, ev.y) == '#1':
+            item = self.vars.identify('item', ev.x, ev.y)
+        if item and item in self.var_values:
+            return(self.var_values[item], ev.x + self.vars.winfo_rootx() + 10,
+                                          ev.y + self.vars.winfo_rooty() + 5)
+        return None
+
     def sync_source_line(self):
         frame = self.frame
         if not frame:
             return
         filename, lineno = self.__frame2fileline(frame)
         if filename[:1] + filename[-1:] != "<>" and os.path.exists(filename):
-            self.flist.gotofileline(filename, lineno)
+            if self.var_open_source_windows.get() or\
+                                    self.flist.already_open(filename):
```

Comment:
```suggestion
            if self.var_open_source_windows.get() or \
               self.flist.already_open(filename):
```

Reply:
The space is not needed, but I prefer to avoid \ at line ends.
```suggestion
            if (self.var_open_source_windows.get() or
                self.flist.already_open(filename)):
```